File: Week06/LeetCode-0973-1.cpp
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        // O(nlogn) algorithm with O(n) extra space
        vector<pair<double, int>> dists;
        for (int i=0; i<points.size(); i++){
            double dist = sqrt(points[i][0]*points[i][0] +
                               points[i][1]*points[i][1]);

            dists.push_back({dist, i});
        } // end-for
        
        // Sort in increasing order
        std::sort(dists.begin(), dists.end());
        
        // Take the first "k" points
        vector<vector<int>> result;
        for (int i=0; i<k; i++)
            result.push_back(points[dists[i].second]);
        return result;
    } // end-kClosest
};

File: Week06/LeetCode-0973-1.py
import heapq
from math import sqrt
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        # O(nlogn) algorithm with O(n) extra space
        dists = []
        for i in range(len(points)):
            dist = sqrt(points[i][0]*points[i][0] + points[i][1]*points[i][1])
            dists.append([dist, i])
            
        # Sort in increasing order
        dists.sort()
        
        # Take the first "k" points
        result = []
        for i in range(k):
            result.append(points[dists[i][1]])
        return result
